src/ndarray/ndarray.cc

```diff
@@ -64,17 +158,147 @@ nnvm::Symbol NDArray::get_autograd_symbol() const {
+  for (size_t i = 0; i < dims.size(); i++)
+    dims[i] = pd.desc().data.dims[i];
+  mkldnn::memory::format cpp_format = static_cast<mkldnn::memory::format>(format);
+  mkldnn::memory::data_type cpp_type = static_cast<mkldnn::memory::data_type>(
+      pd.desc().data.data_type);
+  mkldnn::memory::desc data_md(dims, cpp_type, cpp_format);
+  mkldnn_mem_ptr def_mem(new mkldnn::memory(mkldnn::memory::primitive_desc(data_md,
+          pd.get_engine())));
+
+  MKLDNNStream &stream = MKLDNNStream::Instance();
+  stream.RegisterMem(mem);
+  stream.RegisterMem(def_mem);
+  stream.RegisterPrim(mkldnn::reorder(*mem, *def_mem));
+  if (submit_now)
+    stream.Submit();
+  return def_mem;
+}
+
+NDArray NDArray::ReshapeMKLDNN(const TShape &shape) const {
+  CHECK(!is_none()) << "NDArray is not initialized";
+  CHECK_GE(shape_.Size(), shape.Size())
+    << "NDArray.Reshape: target shape size is larger current shape";
+  if (storage_type() == kDefaultStorage) {
+    NDArray ret = this->Detach();
+    ret.shape_ = shape;
+    return ret;
+  } else if (storage_type() == kMKLDNNStorage) {
+    NDArray ret(kMKLDNNStorage, shape, ctx(), ptr_->delay_alloc, dtype());
+    CHECK(ptr_->Mkl_mem_ != nullptr);
+    // We shouldn't submit the reorder primitive here because submit will
+    // be called in operators.
+    ret.ptr_->Mkl_mem_ = Reorder2Default(ptr_->Mkl_mem_, false);
+    return ret;
+  }
+  LOG(FATAL) << "Reshape for storage type " << storage_type() << " is not implemented yet";
+  return NDArray();
+}
+
+#endif
+
 NDArray NDArray::Reshape(const TShape &shape) const {
   CHECK(!is_none()) << "NDArray is not initialized";
-  auto stype = storage_type();
-  // reshape is not supported for non-default ndarray with dismatching shapes
-  CHECK((shape_ == shape) || stype == kDefaultStorage)
-    << "Reshape for storage type " << stype << " is not implemented yet";
   CHECK_GE(shape_.Size(), shape.Size())
     << "NDArray.Reshape: target shape size is larger current shape";
-  NDArray ret = this->Detach();
-  ret.shape_ = shape;
-  return ret;
+  if (storage_type() == kDefaultStorage) {
+    NDArray ret = this->Detach();
+    ret.shape_ = shape;
+    return ret;
+#if MXNET_USE_MKLDNN == 1
+  } else if (storage_type() == kMKLDNNStorage) {
+    NDArray ret(kMKLDNNStorage, shape, ctx(), ptr_->delay_alloc, dtype());
+    // We need to convert the MKL memory to the default layout.
+    Engine::Get()->PushSync([&](RunContext ctx) {
+        if (this->ptr_->Mkl_mem_) {
+          auto def_format = GetDefaultFormat(this->ptr_->Mkl_mem_->get_primitive_desc().desc());
+          if (this->ptr_->Mkl_mem_->get_primitive_desc().desc().data.format != def_format) {
+            ret.ptr_->Mkl_mem_ = Reorder2Default(this->ptr_->Mkl_mem_);
+          } else {
+            ret.ptr_->Mkl_mem_ = this->ptr_->Mkl_mem_;
+          }
+        }
+    }, ctx(), {this->var()}, {ret.var()},
+    FnProperty::kNormal, 0, PROFILER_MESSAGE("SyncMKLDNN2Default"));
+    ret.WaitToRead();
```

Comment:
No need to sync?

Reply:
I tried not using WaitToRead and saw some non-deterministic memory error. I'm not entirely sure how the async exec engine works. I wanted to ask you what happens if WaitToRead() isn't called here.

Reply:
are you capturing anything by reference?